File: Problems/No.2461_Maximum_Sum_of_Distinct_Subarrays_With_Length_K/maximum_sum_of_distinct_subarrays_with_length_k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        
        // parameters
        long long temp_sum = 0;
        long long MAX_sum = 0;
        std::unordered_map<int, int> distinct_num_count;

        // start computing
        for (int i=0; i<nums.size(); i++) {

            distinct_num_count[nums[i]]++;
            temp_sum += nums[i];
            
            if (i < k-1) continue;
            else {

                if (distinct_num_count.size() == k)
                    MAX_sum = std::max(MAX_sum, temp_sum);
                
                int substract = nums[i-k+1];
                temp_sum -= substract;

                if (distinct_num_count[substract] == 1) distinct_num_count.erase(substract);
                else distinct_num_count[substract]--;

            }

        }

        return MAX_sum;

    }
};

File: Problems/No.2461_Maximum_Sum_of_Distinct_Subarrays_With_Length_K/maximum_sum_of_distinct_subarrays_with_length_k.py
class Solution(object):
    def maximumSubarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """

        # parameters
        temp_sum, MAX_sum = 0, 0
        distinct_num_count = {}

        # start computing
        for i in range(len(nums)):
            distinct_num_count[nums[i]] = distinct_num_count.get(nums[i], 0) + 1
            temp_sum += nums[i]

            if (i < k-1):
                continue
            else:
                if (len(distinct_num_count) == k):
                    MAX_sum = max(MAX_sum, temp_sum)
                
                substract = nums[i-k+1]
                temp_sum -= substract

                if (distinct_num_count[substract] == 1):
                    del distinct_num_count[substract]
                else:
                    distinct_num_count[substract] -= 1
        
        return MAX_sum
